### DISCUSSION TITLE
Are you supposed to return from Next.js API routes?

### DISCUSSION THREAD

**Discussion by IDrumsey** [vercel/next.js#60516] [Help] [answered]
### Summary

When looking at the [docs](https://nextjs.org/docs/pages/building-your-application/routing/api-routes) for API routes using the pages router, it doesn't show returning the `res.status.json`. The problem is when you go to send the response early in your API routes and don't return, all the stuff after runs too. Maybe it's on the docs somewhere and I just haven't seen it. I think it's pretty safe to assume returning in some capacity is safe in API routes to avoid this issue, but want to confirm as not seeing it anywhere on those docs is a little sus.

Example - The problem
```typescript
import type { NextApiRequest, NextApiResponse } from 'next'

type Data = {
  name: string
}

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<Data>
) {

  // want to return the response here, but without "return res.status..." the api route just keeps going
  res.status(200).json({ name: 'John Doe' })

  // this will execute and I don't want it to
  console.log("Doing some other stuff")
}

```

So is the following ok in Next.js API routes?
```typescript
import type { NextApiRequest, NextApiResponse } from 'next'

type Data = {
  name: string
}

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<Data>
) {

  // now the return causes the api route to stop
  return res.status(200).json({ name: 'John Doe' })

  // this won't execute 👍
  console.log("Doing some other stuff")
}

```

**Comment by icyJoseph** [comment, marked as answer]
It is ok to return, but it is not necessary.

```ts
import type { NextApiRequest, NextApiResponse } from 'next'

type Data = {
  name: string
}

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<Data>
) {
  // this is a bad approach
  if (someCondition) { 
      res.status(200).json({ msg: 'msg' });
  }
  // now the return causes the api route to stop
  res.status(200).json({ name: 'John Doe' })
}
```

In the above, for example, the first call would resolve the handler, and the second would be redundant. You could either, do `if-else`, or return in the condition, and you could even do it in two ways:

```ts
  if (someCondition) { 
      res.status(200).json({ msg: 'msg' });
      return;
  }
```

or 

```ts
  if (someCondition) { 
      return res.status(200).json({ msg: 'msg' });
  }
```

These functions, `.json` and others,  return void, and their result is not used by the framework.

One way to think about this is also, promises:

```js
await new Promise((resolve) => {
    resolve(5);
    console.log('I still execute');
})
```

You can think that, certain methods in the `res` object, when called, resolve the promise, which sends the actual response to the client, but things could still happen afterwards.

**Reply by IDrumsey** [reply]
Awesome, thank you!
